### PR TITLE
[#4476] Feat: add token bucket ratelimit filter

### PR DESCRIPTION
**Pull request by anoymouscoder**

### CODE REVIEW

**Review by flycash — COMMENTED** [review]
Could you provide the code of `rate limit filter` ? I want to know how users use those algorithms

**Comment by anoymouscoder** [comment]
How to use:
```
// Construct limiter with options such as rate and capacity.
opts := []LimiterOption{WithRate(1 * time.Millisecond), WithCapacity(1)}
limiter := NewLimiter(opts)
// Register limiter to specify route 
err := handler.InsertFilter("*", web.BeforeRouter, limiter)
```

File: server/web/filter/ratelimit/token_bucket.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+type tokenBucket struct {
+	sync.RWMutex
+	remaining   uint
+	capacity    uint
+	lastCheckAt time.Time
+	rate        time.Duration
+}
+
+// NewTokenBucket return an bucket that implements token bucket
+func NewTokenBucket(opts []BucketOption) Bucket {
```
Comment (flycash) [inline server/web/filter/ratelimit/token_bucket.go:17]: Using `opts ... BucketOption`. In general, when we use `option design pattern`, we always use `opts...XXXXOpt`

**Review by flycash — CHANGES_REQUESTED** [review]
Good!

File: server/web/filter/ratelimit/limiter.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	bucket := l.getBucket(r)
+	if bucket == nil {
+		return true
+	}
+	return bucket.Take(amount)
+}
+
+func (l *limiter) getBucket(r *http.Request) Bucket {
+	key := l.sessionKey(r)
+	b, ok := l.buckets[key]
```
[the remark this answers is not in the document]
Comment (anoymouscoder) [reply server/web/filter/ratelimit/limiter.go:132]: Why is a read lock needed here?

**Review by flycash — APPROVED** [review]
Please resolve the conflict

**Comment by anoymouscoder** [comment]
@flycash Conflicts resolved.

**Pull request merged**